dowhy/gcm/influence.py

```diff
@@ -0,0 +1,339 @@
+    input_samples = shape_into_2d(input_samples)
+
+    if input_subsets is None:
+        input_subsets = [[i] for i in range(input_samples.shape[1])]
+
+    if difference_estimation_func is None:
+        if isinstance(conditional_stochastic_model, ProbabilityEstimatorModel):
+            difference_estimation_func = estimate_kl_divergence_of_probabilities
+        else:
+            def difference_estimation_func(old, new):
+                return np.var(new) - np.var(old)
+
+    if isinstance(conditional_stochastic_model, ProbabilityEstimatorModel):
+        samples_creation_method = conditional_stochastic_model.estimate_probabilities
+    else:
+        samples_creation_method = conditional_stochastic_model.draw_samples
+
+    with warnings.catch_warnings():
+        warnings.filterwarnings('ignore')
+
+        def parallel_job(subset: List[int],
+                         parallel_random_seed: int):
+            set_random_seed(parallel_random_seed)
+
+            return _estimate_direct_strength(
+                samples_creation_method,
+                input_samples,
+                subset,
+                difference_estimation_func,
+                num_samples_from_conditional,
+                max_num_runs,
+                tolerance)
+
+        random_seeds = np.random.randint(np.iinfo(np.int32).max, size=len(input_subsets))
+        results = Parallel(n_jobs=n_jobs)(delayed(parallel_job)(subset, random_seed)
+                                          for subset, random_seed in zip(input_subsets, random_seeds))
+
+    if np.any(results == np.inf):
+        _logger.warning('At least one arrow strength is infinite. This typically happens if the causal models are '
+                        'deterministic, i.e. there is no noise or it is extremely small.')
+
+    return np.array(results)
+
+
+def _estimate_direct_strength(draw_samples_func: Callable[[np.ndarray], np.ndarray],
+                              distribution_samples: np.ndarray,
+                              target_columns: List[int],
+                              difference_estimation_func:
+                              Callable[[np.ndarray, np.ndarray], Union[np.ndarray, float]],
+                              num_samples_conditional: int,
+                              max_num_runs: int,
+                              tolerance: float) -> float:
+    distribution_samples = shape_into_2d(distribution_samples)
+
+    aggregated_conditional_difference_result = 0
+    average_difference_result = 0
+    converged_run = 0
+    for run, sample in enumerate(distribution_samples):
+        tmp_samples = repmat(sample, num_samples_conditional, 1)
+
+        rnd_permutation = np.random.choice(distribution_samples.shape[0],
+                                           num_samples_conditional,
+                                           replace=False)
+
+        removed_arrow_sampling = shape_into_2d(distribution_samples[:, target_columns][rnd_permutation])
```

Comment:
does this mean that the values of target_columns are randomized? you may consider making the variable names less confusing: earlier target_node was the main node with parents. Here `target_columns` refer to the parents of the target node. (thought this is not necessary)

Reply:
I refactored the variable names, it was indeed confusing. The ```target_columns``` here is referring to the subset of edges we are interested in, which is equivalent to the parent column IDs. I also added some inline comments.